runners/datafusion-rust/src/main.rs

```diff
@@ -0,0 +1,291 @@
+    }
+
+    Ok(())
+}
+
+pub async fn execute_query(
+    ctx: &SessionContext,
+    query_path: &str,
+    query_no: u8,
+    debug: bool,
+    output_path: &str,
+    iterations: u8,
+    results: &mut Results,
+) -> Result<()> {
+    let filename = format!("{}/q{query_no}.sql", query_path);
+    println!("Executing query {} from {}", query_no, filename);
+    let sql = fs::read_to_string(&filename)?;
+
+    // some queries have multiple statements
+    let sql = sql
+        .split(';')
+        .filter(|s| !s.trim().is_empty())
+        .collect::<Vec<_>>();
+
+    let multipart = sql.len() > 1;
+
+    let mut durations = vec![];
+    for iteration in 0..iterations {
+        // duration for executing all queries in the file
+        let mut total_duration_millis = 0;
+
+        for (i, sql) in sql.iter().enumerate() {
+            if debug {
+                println!("Query {}: {}", query_no, sql);
+            }
+
+            let file_suffix = if multipart {
+                format!("_part_{}", i + 1)
+            } else {
+                "".to_owned()
+            };
+
+            let start = Instant::now();
+            let df = ctx.sql(sql).await?;
+            let batches = df.clone().collect().await?;
+            let duration = start.elapsed();
+            total_duration_millis += duration.as_millis();
+
+            let row_count: usize = batches.iter().map(|b| b.num_rows()).sum();
+
+            println!(
+                "Query {query_no}{file_suffix} executed in: {duration:?} and returned {row_count} rows",
+            );
+
+            if iteration == 0 {
+                let plan = df.clone().into_optimized_plan()?;
+                let formatted_query_plan = format!("{}", plan.display_indent());
+                if debug {
+                    println!("{}", formatted_query_plan);
+                }
+
+                let exec = df.create_physical_plan().await?;
+                if debug {
+                    println!("{}", displayable(exec.as_ref()).indent(false));
+                }
+
+                let filename = format!(
+                    "{}/q{}{}_logical_plan.txt",
+                    output_path, query_no, file_suffix
+                );
+                let mut file = File::create(&filename)?;
+                write!(file, "{}", formatted_query_plan)?;
+
+                // write results to disk
+                if batches.is_empty() {
+                    println!("Empty result set returned");
+                } else {
+                    let filename = format!("{}/q{}{}.csv", output_path, query_no, file_suffix);
+                    let t = MemTable::try_new(batches[0].schema(), vec![batches])?;
+                    let df = ctx.read_table(Arc::new(t))?;
+                    df.write_csv(&filename, DataFrameWriteOptions::default(), None)
+                        .await?;
+                }
+            }
+        }
+        durations.push(total_duration_millis);
```

Comment:
Looks like this takes total duration of all iterations. I'm wondering if we should use average duration?

Reply:
I think this is just hard to parse from looking at the diff here, but the code is:

```rust
    for iteration in 0..iterations {
        // duration for executing all queries in the file
        let mut total_duration_millis = 0;
        for (i, sql) in sql.iter().enumerate() {
            ...
        }
        durations.push(total_duration_millis);
    }
```

Reply:
`total_duration_millis` is the total duration for all queries in the file (some query files have multiple queries)